expr/func_offset.go

```diff
@@ -0,0 +1,48 @@
+package expr
+
+import (
+	"fmt"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/schema"
+)
+
+type FuncOffset struct {
+	in     GraphiteFunc
+	factor float64
+}
+
+func NewOffset() GraphiteFunc {
+	return &FuncOffset{}
+}
+
+func (s *FuncOffset) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgFloat{key: "factor", val: &s.factor},
+		}, []Arg{
+			ArgSeriesList{},
+		}
+}
+
+func (s *FuncOffset) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncOffset) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
```

Comment:
I'm a little confused, what does `in` do/how does it get populated?

Reply:
this is a big topic. and you'll find some stuff in devdocs.
but TLDR, functions recursively call Exec on their GraphiteFunc inputs, until it hits the last function which is FuncGet, which pulls the data out of a map.
that map has been loaded with data before the function processing runs